src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -396,33 +396,99 @@ def _list_offers_aggregator(self, arguments):
+                self.offers.replace(offer, new_offer, iterated_market.id)
+                return {
+                    "command": "update_offer",
+                    "area_uuid": self.device.uuid,
+                    "status": "ready",
+                    "offer": offer.to_JSON_string(),
+                    "transaction_id": arguments.get("transaction_id", None),
+                }
+            except MarketException:
+                continue
+
+    def _offer_aggregator(self, arguments):
+        assert set(arguments.keys()) == {'price', 'energy', 'transaction_id', 'type'}
+        with self.lock:
+            arguments['seller'] = self.device.name
+            arguments['seller_origin'] = self.device.name
+            try:
+                assert arguments['energy'] <= self.state.energy_to_sell_dict[self.market.time_slot]
+                offer = self.market.offer(
+                    price=arguments['price'], energy=arguments['energy'],
+                    seller=arguments['seller'], seller_origin=arguments['seller_origin']
+                )
+                self.offers.post(offer, self.market.id)
+                self.state.offered_sell_kWh[self.market.time_slot] = \
+                    self.offers.open_offer_energy(self.market.id)
+                self.state.clamp_energy_to_sell_kWh([self.market.time_slot])
+                return {
+                    "command": "offer",
+                    "area_uuid": self.device.uuid,
+                    "status": "ready",
+                    "offer": offer.to_JSON_string(),
+                    "transaction_id": arguments.get("transaction_id", None),
+                }
+            except Exception as e:
+                return {
+                    "command": "offer", "status": "error",
+                    "area_uuid": self.device.uuid,
+                    "error_message": f"Error when handling offer create "
+                                     f"on area {self.device.name} with arguments {arguments}.",
+                    "transaction_id": arguments.get("transaction_id", None)}
+
+    def _update_bid_aggregator(self, arguments):
+        assert set(arguments.keys()) == {'price', 'energy', 'type', 'transaction_id'}
+        bid_rate = arguments["price"] / arguments["energy"]
+        with self.lock:
+            existing_bids = list(self.get_posted_bids(self.market))
+            existing_bid_energy = sum([bid.energy for bid in existing_bids])
+
+            for bid in existing_bids:
+                assert bid.buyer == self.owner.name
+                if bid.id in self.market.bids.keys():
+                    bid = self.market.bids[bid.id]
+                self.market.delete_bid(bid.id)
+
+                self.remove_bid_from_pending(self.market.id, bid.id)
+            if len(existing_bids) > 0:
+                updated_bid = self.post_bid(self.market, bid_rate * existing_bid_energy,
+                                            existing_bid_energy, buyer_origin=self.device.name)
```

Comment:
Ok, i just realise that you also combining multiple old bids into one new one. I definitely not agree with this. But as Colin wants it like this. Leave it.

Reply:
I agree with your reservations but for now his api-script is also placing one offer/bid at a time, so its okay for now.